### ISSUE TITLE
==0.1.* doesn't match 0.1+upstream.2

### ISSUE THREAD

**Issue by habnabit** [pypa/packaging#41]
As in the title, `==0.1.*` doesn't match the version `0.1+upstream.2`. I discovered this as a result of specifying a requirement `~=0.1.0`, and getting `0.1` instead of `0.1+upstream.2`. I can produce a reproducing example if necessary.

**Comment by dstufft** [comment]
Some notes from when I was testing this:

```python
>> import packaging.specifiers
>>> packaging.specifiers.SpecifierSet("~=0.1.0").contains("0.1+upstream.2")
False
>>> packaging.specifiers.SpecifierSet(">=0.1.0,==0.1.*").contains("0.1+upstream.2")
False
>>> packaging.specifiers.SpecifierSet(">=0.1.0").contains("0.1+upstream.2")
True
>>> packaging.specifiers.SpecifierSet("==0.1.*").contains("0.1+upstream.2")
False
>>> packaging.specifiers.SpecifierSet("==0.*").contains("0.1+upstream.2")
True
>>> packaging.specifiers.SpecifierSet("==0.1.*").contains("0.1+upstream.2")
False
>>> packaging.specifiers.SpecifierSet("==0.1.0.*").contains("0.1+upstream.2")
False
```

It appears as if we do this correctly when ``==0.*`` is the version specifier, but not when ``==0.1.*`` is the version specifier.

We might also want to double check PEP 440 to make sure that this particular use case is clearly defined.

**Comment by dstufft** [comment]
Oh, and for completeness sake:

```python
>>> packaging.specifiers.SpecifierSet("==0.1.*").contains("0.1")
True
```

**Comment by habnabit** [comment]
While I'm thinking about this, it would also be nice to be able to have requirements be able to match against upstream versions. `>=0.1.0+upstream.2` ignores the bit after the `+`, and effectively works the same as `>=0.1.0`.

**Comment by dstufft** [comment]
Double checking PEP 440, I see why that doesn't work:

> *[PEP 440](https://www.python.org/dev/peps/pep-0440/#inclusive-ordered-comparison) says:*
> Local version identifiers are NOT permitted in this version specifier.

I don't recall exactly why we did that, I think it was likely to prevent (or discourage) people from uploading things to PyPI that had a dependency on a particular local version. In fact, the PEP disallows local versions like that in all specifiers except for ``==`` (without the ``.*``) and ``===``. The "ignoring" behavior you're seeing is because we fall back onto a legacy implementation if something isn't a valid PEP 440 specifier which I think probably has that behavior.

**Comment by dstufft** [comment]
Oh, to be clear, changing the behavior of ``>=`` will require at a minimum, discussion on distutils-sig, but the behavior of ``~=`` is a bug.

**Comment by xavfernandez** [comment]
The `~=` behavior should be fixed by #47
